### PR TITLE
[portsorch]: Add support of cable breakout feature

### PR DESCRIPTION
**Pull request by volodymyrsamotiy**

### CODE REVIEW

File: orchagent/portsorch.cpp [hunk 1]
```diff
@@ -461,6 +461,107 @@ void PortsOrch::updateDbPortOperStatus(sai_object_id_t id, sai_port_oper_status_
     }
 }
 
+bool PortsOrch::createPort(set<int> lane_set, uint32_t speed)
```
Comment (oleksandrivantsiv) [inline orchagent/portsorch.cpp]: It is better to pass parameter by refference: const set<int>& lane_set

File: orchagent/portsorch.cpp [hunk 2]
```diff
@@ -461,6 +461,107 @@ void PortsOrch::updateDbPortOperStatus(sai_object_id_t id, sai_port_oper_status_
...
+    }
+
+    m_portListLaneMap[lane_set] = port_id;
+
+    SWSS_LOG_INFO("Create port %lx with the speed %u", port_id, speed);
+
+    return true;
+}
+
+bool PortsOrch::removePort(set<int> lane_set)
```
Comment (oleksandrivantsiv) [inline orchagent/portsorch.cpp]: Same here

File: orchagent/portsorch.cpp [hunk 3]
```diff
@@ -461,6 +461,107 @@ void PortsOrch::updateDbPortOperStatus(sai_object_id_t id, sai_port_oper_status_
...
+        SWSS_LOG_ERROR("Failed to remove port %lx, rv:%d", m_portListLaneMap[lane_set], status);
+        return false;
+    }
+
+    SWSS_LOG_INFO("Remove port %lx", m_portListLaneMap[lane_set]);
+
+    return true;
+}
+
+bool PortsOrch::initPort(string alias, set<int> lane_set)
```
Comment (oleksandrivantsiv) [inline orchagent/portsorch.cpp]: And here

Comment (oleksandrivantsiv) [inline orchagent/portsorch.cpp]: Also for string: const string&

File: orchagent/portsorch.cpp [hunk 4]
```diff
@@ -461,6 +461,107 @@ void PortsOrch::updateDbPortOperStatus(sai_object_id_t id, sai_port_oper_status_
...
+    SWSS_LOG_INFO("Create port %lx with the speed %u", port_id, speed);
+
+    return true;
+}
+
+bool PortsOrch::removePort(set<int> lane_set)
+{
+    SWSS_LOG_ENTER();
+
+    sai_status_t status = sai_port_api->remove_port(m_portListLaneMap[lane_set]);
```
Comment (oleksandrivantsiv) [inline orchagent/portsorch.cpp]: Are you sure that m_portListLaneMap always contains lane set?

File: orchagent/portsorch.cpp [hunk 5]
```diff
@@ -530,47 +643,52 @@ void PortsOrch::doPortTask(Consumer &consumer)
...
-
-                        p.m_index = (uint32_t)m_portList.size(); // TODO: Assume no deletion of physical port
-                        p.m_port_id = id;
+                        ++it;
+                    }
+                }
 
-                        /* Initialize the port and create router interface and host interface */
-                        if (initializePort(p))
+                for (map<set<int>, tuple<string, uint32_t>>::iterator it = m_lanesALiasSpeedMap.begin(); it != m_lanesALiasSpeedMap.end();)
```
Comment (oleksandrivantsiv) [inline orchagent/portsorch.cpp]: Please add typedef for map<set<int>, tuple<string, uint32_t>> type or use foreach and auto instead for loop

**Comment by stcheng** [comment]
hello @volodymyrsamotiy , i have removed the branch v1.0.1 and rename it to v1.0.3. Could you generate a new pull request?
Another question is that is this pull request only applied to SAI 1.0?

**Comment by volodymyrsamotiy** [comment]
Hi @stcheng,
1. I have created new pull request for the v1.0.3 branch.
2. Yes, it is for SAI v1.0.

**Pull request closed without merging**